stout/take.h

```diff
@@ -0,0 +1,191 @@
+#include <vector>
+
+#include "stout/eventual.h"
+#include "stout/filter.h"
+#include "stout/stream.h"
+
+namespace stout {
+namespace eventuals {
+namespace detail {
+
+////////////////////////////////////////////////////////////////////////
+
+struct _TakeLastN {
+  template <typename K_, typename Arg_>
+  struct Continuation : public TypeErasedStream {
+    // NOTE: explicit constructor because inheriting 'TypeErasedStream'.
+    Continuation(K_ k, size_t n)
+      : k_(std::move(k)),
+        N_(n) {}
+
+    void Start(TypeErasedStream& stream) {
+      stream_ = &stream;
+      data_ = std::vector<int>(N_);
+      k_.Start(*this);
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    template <typename... Args>
+    void Body(Args&&... args) {
+      if (i_ == data_.size()) {
+        i_ = 0;
+        filled_ = true;
+      }
+      if (filled_) {
+        for (size_t j = 0; j < data_.size() - 1; ++j) {
+          data_[j] = data_[j + 1];
+        }
+        SetValue(data_.size() - 1, std::forward<Args>(args)...);
+      } else {
+        SetValue(i_++, std::forward<Args>(args)...);
+      }
+      stream_->Next();
+    }
+
+    void Ended() {
+      if (endedTriggered_) {
+        k_.Ended();
```

Comment:
Under what circumstances is `Ended()` called twice on `_TakeLastN`? If never, can we remove this `if`-statement?